### ISSUE TITLE
Space between text and underline

### ISSUE THREAD

**Issue by felangga** [flutter/flutter#30541]
I'm using Text.rich with underlined option to make underlined text. It works, but it's possible to add space between the word and underline line? Or make the line more bolder ?

**Comment by IvanTanaka** [comment]
Is it possible to add spacing between  Text and it's underline by using TextDecoration.underline in TextStyle?

**Comment by cadinz** [comment]
this is not great but it works.. 

Column(
                          children: <Widget>[
                            FittedBox(
                              child: Container(
                                child: Stack(
                                  children: <Widget>[
                                    Padding(
                                      padding: const EdgeInsets.only(bottom: 3),
                                      child: Row(
                                        crossAxisAlignment: CrossAxisAlignment.start,
                                        mainAxisAlignment: MainAxisAlignment.center,
                                        children: <Widget>[
                                          Text(
                                            'hello',
                                            style: TextStyle(
                                                fontFamily: 'nanum',
                                                color: Colors.white,
                                                fontSize:
                                                    ScreenUtil.getInstance().setSp(28)),
                                          ),
                                          Text(
                                            'world',
                                            style: TextStyle(
                                                fontFamily: 'nanum',
                                                fontWeight: FontWeight.w700,
                                                color: Colors.white,
                                                fontSize:
                                                    ScreenUtil.getInstance().setSp(28)),
                                          ),
                                        ],
                                      ),
                                    ),
                                    Positioned.fill(
                                      child: Column(
                                        mainAxisAlignment: MainAxisAlignment.end,
                                        children: <Widget>[
                                          SizedBox(height:5),
                                          Container(
                                            height: 1,
                                            color: Colors.white,
                                          ),
                                        ],
                                      ),
                                    ),
                                  ],
                                ),
                              ),
                            ),

                          ],
                        ),

**Comment by ozc** [comment]
`
Container(
                    padding: EdgeInsets.only(
                      bottom: 3, // space between underline and text
                    ),
                    decoration: BoxDecoration(
                      border: Border(bottom: BorderSide(
                        color: Colors.white,  // Text colour here
                        width: 1.0, // Underline width
                      ))
                    ),

                    child: Text(
                      "Text Here",
                      style: TextStyle(
 color: Colors.white,  // Text colour here
                      ),
                    ),
                  )
`

**Comment by pr-1** [comment]
> I'm using Text.rich with underlined option to make underlined text. It works, but it's possible to add space between the word and underline line? Or make the line more bolder ?

You can Use the decorationThickness: property on Text to make it bolder. But the spacing problem is not yet fixed i suppose. A workaround is what @ozc told.

**Comment by jlubeck** [comment]
Any updates?

**Comment by jtmuller5** [comment]
I came up with this hack that uses just the Text widget. In short, you hide the actual text using Colors.transparent and offset it's shadow so it floats above the underline as much as you want.

```
Text(
            "Forgot Password?",
            style: TextStyle(
              shadows: [
                Shadow(
                    color: Colors.black,
                    offset: Offset(0, -5))
              ],
              color: Colors.transparent,
              decoration:
              TextDecoration.underline,
              decorationColor: Colors.blue,
              decorationThickness: 4,
              decorationStyle:
              TextDecorationStyle.dashed,
            ),
          )
```

**Comment by guilhermepata** [comment]
Any updates on this? For me to use the container box decoration workaround I have to use a `WidgetSpan` inside a `Text.rich`, since I'm trying to add links to a body of text. This forces me to set the forceStructHeight parameter to true, or else the line spacing will increase to fit the container's border. However, links at the last line in a body of text will have their underlines cropped...

**Comment by guilhermepata** [comment]
Update! Found a solution that sort of combines @cadinz and @ozc's methods: A `Stack` whose children are the desired `Text` and a `Positioned`, the latter wrapping a `Container` whose border is the underline. Since the `Positioned`'s `top` depends on the `fontSize` and `height` of our `TextStyle,` this widget can be wrapped in a `WidgetSpan` inside a `Text.rich` (or `RichText`), and we can set `forceStrutHeight` to `true` without clipping the underline. If the underline is clipped (which can happen on large font sizes), we can simply adjust the `top` parameter accordingly.
```dart
class UnderlinedText extends StatelessWidget {
  final String text; // text to underline
  final Color decorationColor; // underline color
  final double top;
  final TextStyle style;

  const UnderlinedText(this.text,
      {Key key, this.decorationColor, this.top, @required this.style})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Text(  // text widget that will be seen
          text,
          style: style,
          maxLines: 1, 
        ),
        Positioned(
            top: top ?? // adjust if necessary
                style.height *
                    style.fontSize *
                    .96, // distance based on font size
            child: Container(
              height: 0,
              decoration: BoxDecoration(
                border: Border(
                  bottom: BorderSide(
                    color: decorationColor, // Decoration color here
                    width: 1.5, // Underline width
                  ),
                ),
              ),
              child: Text( // invisible text widget, needed to set the container's size
                text,
                style: style.copyWith(
                    color:
                        Colors.transparent), // this one has to be transparent
                maxLines: 1,
              ),
            )),
      ],
    );
  }
}
```

**Comment by joanofdart** [comment]
@jtmuller5 's solution is quite amazing, mad props to you sir.

**Comment by farhanfarz** [comment]
Inspired from answer by @jtmuller5, This workaround should do if you wish to underline only a part of the text:



`Color kTitleColor = Colors.black;

Widget buildAcceptTerms(BuildContext context) {
    TextStyle defaultStyle = TextStyle(
      color: Colors.transparent,
      decorationColor: kTitleColor,
      fontSize: 15.0,
      shadows: [Shadow(color: kTitleColor, offset: Offset(0, -5))],
    );
    TextStyle linkStyle = TextStyle(
        color: Colors.transparent,
        decorationColor: kTitleColor,
        decoration: TextDecoration.underline,
        decorationThickness: 2.0,
        shadows: [Shadow(color: kTitleColor, offset: Offset(0, -5))],
        fontSize: 15.0);
    return RichText(
      text: TextSpan(
        style: defaultStyle,
        children: <TextSpan>[
          TextSpan(text: 'Accept the '),
          TextSpan(
            text: 'terms and conditions',
            style: linkStyle,
            recognizer: TapGestureRecognizer()
              ..onTap = () {
                //showTermsSheet(context);
              },
          ),
          TextSpan(text: ' of this transfer'),
        ],
      ),
    );
  }`

**Comment by freitzzz** [comment]
Updates on this? :(

**Comment by MufakirMairaj** [comment]
> ```
> Container(
>                     padding: EdgeInsets.only(
>                       bottom: 3, // space between underline and text
>                     ),
>                     decoration: BoxDecoration(
>                       border: Border(bottom: BorderSide(
>                         color: Colors.white,  // Text colour here
>                         width: 1.0, // Underline width
>                       ))
>                     ),
> 
>                     child: Text(
>                       "Text Here",
>                       style: TextStyle(
>  color: Colors.white,  // Text colour here
>                       ),
>                     ),
>                   )
> ```

Yes,
![New Project](https://user-images.githubusercontent.com/52829267/124113759-3df67680-da85-11eb-90e0-8182b915d3f4.png)
 it is good for one-line text what about multiple lines in Text Widget in flutter.

**Comment by Cedricoh** [comment]
Bumping this issue, I would really appreciate seeing this implemented. Thanks

**Comment by mr-mmmmore** [comment]
Bumping again because although the various solutions proposed here work they are only hacks to an inadequate behavior:  the underline shouldn't stick to the text.

So adding an option to control the space over the underline or simply moving the line a little bit away from the text would be great.

**Comment by mr-mmmmore** [comment]
In the wait for a solution I have created an extension to underline a text easily using the shadow hack but without cluttering the widget's code. It's used like this:

```dart
Widget build(BuildContext context) {
  final myStyle = Theme.of(context).textTheme.headline4;

  return Text(
    'Hello, World!',
    //------------------------------------------------
    // simply apply the underlined method to the style
    style: myStyle?.underlined(distance: 2),
    //------------------------------------------------
  );
}
```
The complete code is [here](https://stackoverflow.com/a/69343754).

**Comment by EvGen94** [comment]
I hope a solution will be added soon to the framework

**Comment by roked** [comment]
Up!

**Comment by aytunch** [comment]
The workarounds stated above can not be used for multiline `RichText` where we only want to underline a specific word. Setting the underline offset should not be something done with a workaround.

**Comment by UP929312** [comment]
Bump

**Comment by justinschier** [comment]
Bump

**Comment by AlexV525** [comment]
cc @GaryQian

**Comment by DavidOrakpo** [comment]
2022 october, no update. Wow

**Comment by FlashTang** [comment]
2022 Nov, no update. Wow

**Comment by TheCarpetMerchant** [comment]
I'm again bumping this. Flutter is a UI framework, the fact that we can't specify the spacing between a text and its underline is baffling.
In the case that this is because of a blocking issue in Skia, please at least let us know and link the related information. @mit-mit

**Comment by keyfayaz** [comment]
Looking for this.

**Comment by fahmidul** [comment]
Here is a solution provided by an individual.
`Text('.....',
                                  style: TextStyle(
                                    color: Colors.transparent,
                                    shadows: [Shadow(color: Colors.black, offset: Offset(0, -3))],
                                    decoration: TextDecoration.underline,
                                    decorationColor: Colors.black,
                                    decorationThickness: 1,

                                  ))`
https://medium.com/@zeerockyf5/how-to-easily-set-space-between-flutter-text-and-underline-89e0333d3d17

**Comment by gguijarro-c-chwy** [comment]
Bump

**Comment by demoYang** [comment]
2023-6-16

**Comment by MarceloRobert** [comment]
I've tried to tackle this issue as not only I wanted a custom space, but also wanted the underline to be painted _under_ the text. My solution was to use a WidgetSpan with a Stack of a Text above a CustomPaint.

TLDR.: it works with other spans, has full control over the spacing and underline style, is customizable, but only works for single line text (it will **not** have a line break) and is definitely a hacky workaround, I don't garantee anything.

It is actually simple, but it's complex underneath, being somewhat easy to break. Inside the Text.rich ( TextSpan ( children [, add:

```dart
WidgetSpan(
  alignment: PlaceholderAlignment.baseline,
  baseline: TextBaseline.alphabetic,
  child: CustomUnderlinedText(
  text: "your text here",
    style: pTextStyle, // your TextStyle here, this is my app's constant, it's just a "fontSize: 16"
    underlineSpacing: 5, // the spacing between the text and the underline
    // you can put other parameters to the custom widget
    // to have more control over it, like the underline color or thickness
  ),
),
```

The CustomUnderlinedText is also simple, defined as:

<details>

<summary>CustomUnderlinedText code, Click to expand</summary>

```dart
class CustomUnderlinedText extends StatelessWidget {
  final double underlineSpacing;
  final String text;
  final TextStyle style;

  const CustomUnderlinedText({
    super.key,
    this.underlineSpacing = 0,
    required this.text,
    required this.style,
  });

  /// retrieves the size of a text before actually painting it
  Size getTextSize(String text, TextStyle style) {
    TextPainter auxTextPainter = TextPainter(
      text: TextSpan(text: text, style: style),
      maxLines: 1,
      textDirection: TextDirection.ltr,
    )..layout(minWidth: 0, maxWidth: double.infinity);
    return auxTextPainter.size;
  }

  @override
  Widget build(BuildContext context) {
    Size canvasSize = getTextSize(text, style);
    return SizedBox( // needs SizedBox for the CustomPaint
      width: canvasSize.width * 1.05, // without the +5% it will overflow
      height: canvasSize.height + (underlineSpacing < 0 ? 0 : underlineSpacing), // without the <0 check it will cut the text
      child: Stack(
        children: [
          Positioned(
            child: CustomPaint(
              painter: CustomUnderline(), // this is your underline painter
              size: Size(
                canvasSize.width * 1.05,
                canvasSize.height + underlineSpacing,
              ),
            ),
          ),
          Text(
            text,
            style: style,
            maxLines: 1,
            overflow: TextOverflow.visible, // doesn't seem to make a difference at all
          ),
          // you can add foreground painters here aswell, or anything for the stack, really
        ],
      ),
    );
  }
}
```

</details>

The CustomUnderline (a custom painter) is a bit more complicated, but not really. The fact is that you can change it for whatever you like, but it is more low lever, so instead of just being "TextDecorationStyle.[anything except solid]" you'll have to code the painter to do so. Anyhow, here's the CustomUnderline I defined:

<details>

<summary>CustomUnderline code, Click to expand</summary>

```dart
class CustomUnderline extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    // Define the line coordinates.
    Offset startPoint = Offset(0, size.height);
    Offset endPoint = Offset(size.width, size.height);

    // Create a Paint object to define the line's appearance.
    Paint paint = Paint()
      ..color = Colors.blue
      ..strokeWidth = 5
      ..strokeCap = StrokeCap.round;

    // Draw the line on the canvas.
    canvas.drawLine(startPoint, endPoint, paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    // Return false to prevent unnecessary repaints.
    return false;
  }
}
```

</details>

Before you leave, be aware of the pros and cons:

**Pros:**

- Somewhat solves your problem <sup>1</sup>
- Paints the underline behind the text, not over it
- Doesn't push the text upwards like the shadow method does
- Can paint custom types of underline, like diagonals or curves

**Cons:**

- <sup>1</sup> Only works for words/phrases with no line break
- Finnicky to get the canvas size right, may need a space after the underlined text
- Possible but maybe hard to make custom underlines

**Considerations:**

- If used in the middle of a paragraph, will push the line under it down (as it should be, in my opinion)
- I did not test with other languages and fonts

If you want to dive deeper into this issue, here's what I found but didn't work (note that I'm not a Flutter expert, so take this with a grain of salt):

One problem I had is that Widgets and Spans are different, so a Stateless/Stateful widget can't return a TextSpan or InlineSpan. Also, the InlineSpan build method doesn't return anything, instead it uses a ParagraphBuilder, which is focused on text and because of that I coudn't find how to paint a CustomPainter there;

The TextDecoration seems to just set a mask for something to build it, but I haven't found where it is used or when it is painted;

The Canvas was somehow an issue, I needed to set a SizedBox in order to paint the line properly, so I don't know how would I paint a line just by using some implementation of a InlineSpan.


I hope this helps, otherwise the transparent text with offset shadow is the best way to make this.
Here's some screenshots of comparisons, the reds are the usual underlines, the greens are the custom ones:

![above-below](https://github.com/flutter/flutter/assets/76781499/bc498625-6785-4b96-bcff-cc97ec2939ea)
![spacing](https://github.com/flutter/flutter/assets/76781499/41b6ebbd-9722-4278-983c-37727c09d5c1)
![styles](https://github.com/flutter/flutter/assets/76781499/331e78e9-0ffa-499f-a9ea-1eeee31a9c03)


And here is an example of it used in a paragraph:

![image](https://github.com/flutter/flutter/assets/76781499/b3161f62-ccb0-42ac-b73c-6d404c74b103)

**Comment by LongCatIsLooong** [comment]
Currently the underline position SkParagraph uses is from the font's [underlinePosition](https://learn.microsoft.com/en-us/typography/opentype/spec/post#header). So for example Roboto (EM size == 2048, ascent == 1536, descent == 512) has an underline position of -200 , in flutter the top of the underline is painted somewhere between the baseline and the bottom of the descender.

I'd like to understand where exactly would you like to place the underline? Is it at the bottom of the em box for each character (so the underline will be placed at the where the descent is), or you would like to override the `underlinePosition` value?

**Comment by TheCarpetMerchant** [comment]
Personally I'd prefer an overwrite.

**Comment by LongCatIsLooong** [comment]
Related: https://github.com/flutter/flutter/issues/42833 where the underline is not aligned for different runs.

**Comment by gnprice** [comment]
Looking at what CSS offers for this for comparison, the [text-underline-position](https://drafts.csswg.org/css-text-decor/#text-underline-position-property) property has two values relevant to horizontal text: `auto` and `under`.
* `auto` is the default and is user-agent-defined, just has to be "at or under the alphabetic baseline". A non-normative note says:
  > It is suggested that the default underline position be close to the alphabetic baseline, unless that would either cross subscripted (or otherwise lowered) text or draw over glyphs from Asian scripts such as Han or Tibetan for which an alphabetic underline is too high: in such cases, shifting the underline lower or aligning to the em box edge as described for [under](https://drafts.csswg.org/css-text-decor/#underline-under) may be more appropriate.

  I'd guess that for Latin and similar scripts, that means browsers end up using the font's underlinePosition metric, just like Flutter does via SkParagraph. Maybe they also have some logic where for scripts like Han they go for the bottom of the em box instead, as this note suggests. (You'd hope that the font would say that too, but fonts don't always do what one would hope).

* `under`. This seems to mean using the bottom of the em box, i.e. at the descent line.

Naturally we don't have to follow CSS's choices here. But it's probably worth considering if we can keep the API from being *more* complicated than CSS's, because in general CSS often goes for making things pretty complicated.

Potentially that points to an enum named with values like:
* `UnderlinePosition.default` — the status quo; possibly in the future gaining some hacks like choosing different positions based on the script; corresponds to CSS `auto`
* `UnderlinePosition.underline` — the font's underlinePosition metric, so also the status quo but explicitly so; possibly needs a clearer name
* `UnderlinePosition.bottom` — the bottom of the em box, corresponding to CSS `under`

I think that would cover the use cases in this thread as far as they've been clearly laid out. It wouldn't cover the screenshots in the very helpful comment https://github.com/flutter/flutter/issues/30541#issuecomment-1656382350 , but those call for a wider scope — they involve underline thickness and color as well as position.

If we then later do want to extend to arbitrary positions, that can be in the form of a `double` that's taken as an offset relative to the line the enum chooses. It's impossible to make sense of a number without *some* baseline anyway, and I think it's likely that someone who's trying to position an underline exactly is often going to want to explicitly choose which baseline to start from. (This might go along with adding more values to the enum, like `alphabeticBaseline`.)

**Comment by gnprice** [comment]
Some of the screenshots in that other thread, like [this one](https://github.com/flutter/flutter/issues/42833#issuecomment-569104680):
![image](https://github.com/flutter/flutter/assets/28173/fae5a96c-ed2d-42e4-9385-9d38719574bd)

also point to it possibly being a good idea to start doing some of those hacks the CSS text alludes to, like seeing the script is Han (or kana, etc.) and putting the underline at the bottom instead of trusting the font's underlinePosition. Looks like that font has a too-high underlinePosition that cuts right through its kanji and kana glyphs.

**Comment by MengLizhi** [comment]
Is there any progress on this issue?

**Comment by demoYang** [comment]
Impeller will support it ?

**Comment by bjornegil** [comment]
On native iOS, the underline breaks on e.g. characters g, j, p and y, which looks much better:

<img width="66" alt="Screenshot 2024-11-06 at 12 31 38" src="https://github.com/user-attachments/assets/3f5fc819-abc1-4158-a901-780b68deaf70">

Would be very nice if that was supported in Flutter, in addition to the offset.

**Comment by silverhairs** [comment]
> I came up with this hack that uses just the Text widget. In short, you hide the actual text using Colors.transparent and offset it's shadow so it floats above the underline as much as you want.
> 
> ```
> Text(
>             "Forgot Password?",
>             style: TextStyle(
>               shadows: [
>                 Shadow(
>                     color: Colors.black,
>                     offset: Offset(0, -5))
>               ],
>               color: Colors.transparent,
>               decoration:
>               TextDecoration.underline,
>               decorationColor: Colors.blue,
>               decorationThickness: 4,
>               decorationStyle:
>               TextDecorationStyle.dashed,
>             ),
>           )
> ```

This is the best workaround so far but unfortunately it doesn't work well with `SelectableText` 😿

**Comment by hamirshekhawat** [comment]
We should be able to use background attribute to create a Paint object that paints the underline. This works great with SelectableText too. Example code, you can change values of offsets based on line height and font size.

<img width="316" alt="Image" src="https://github.com/user-attachments/assets/b1e3fb79-70b9-4a87-84e3-9fe11c02733b" />

`
TextStyle(
  background: Paint()
    ..shader = ui.Gradient.linear(
      const Offset(0, 0),
      const Offset(0, 28),
      [
        Colors.red.shade100,
        Colors.red.shade100,
        Colors.red,
        Colors.red,
      ],
      [
        0.0,
        0.8,
        0.8,
        1.0,
      ],
    ),
  height: 1.55,
);
`

**Comment by rockingdice** [comment]
> shadows: [
>                 Shadow(
>                     color: Colors.black,
>                     offset: Offset(0, -5))
>               ],
>               color: Colors.transparent,
>               decoration:
>               TextDecoration.underline,
>               decorationColor: Colors.blue,
>               decorationThickness: 4,
>               decorationStyle:
>               TextDecorationStyle.dashed,

Brilliant solution!!! You must be a genius.
